src/verifiable/credential.ts

```diff
@@ -134,6 +137,128 @@ export class W3CCredential {
     return undefined;
   }
 
+  /**
+   * Verify credential proof
+   *
+   * @returns {*}  {(boolean)}
+   */
+  async verifyProof(
+    proofType: ProofType,
+    resolverURL: string,
+    credStatusResolverRegistry?: CredentialStatusResolverRegistry
+  ): Promise<boolean> {
+    const proof = this.getProofByType(proofType);
+    if (!proof) {
+      throw new Error('proof not found');
+    }
+
+    const coreClaim = this.getCoreClaimFromProof(proofType);
+    if (!coreClaim) {
+      throw new Error(`can't get core claim`);
+    }
+
+    switch (proofType) {
+      case ProofType.BJJSignature: {
+        if (!credStatusResolverRegistry) {
+          throw new Error('please provide credential status resolver registry');
+        }
+        const userDID = DID.parse(this.credentialSubject.id as string);
+        return this.verifyBJJSignatureProof(
+          proof as BJJSignatureProof2021,
+          coreClaim,
+          resolverURL,
+          userDID,
+          credStatusResolverRegistry
+        );
+      }
+      case ProofType.Iden3SparseMerkleTreeProof: {
+        return this.verifyIden3SparseMerkleTreeProof(
+          proof as Iden3SparseMerkleTreeProof,
+          coreClaim,
+          resolverURL
+        );
+      }
+      default: {
+        throw new Error('invalid proof type');
+      }
+    }
+  }
+
+  private async verifyBJJSignatureProof(
+    proof: BJJSignatureProof2021,
+    coreClaim: Claim,
+    resolverURL: string,
+    userDID: DID,
+    credStatusResolverRegistry: CredentialStatusResolverRegistry
+  ): Promise<boolean> {
+    // issuer auth claim
+    const authClaim = proof.issuerData.authCoreClaim;
+    const rawSlotsInt = authClaim.rawSlotsAsInts();
+    const pubKey = new PublicKey([rawSlotsInt[2], rawSlotsInt[3]]);
+
+    // core claim hash
+    const { hi, hv } = coreClaim.hiHv();
+    const claimHash = poseidon.hash([hi, hv]);
+    const bjjValid = pubKey.verifyPoseidon(claimHash, proof.signature);
+
+    if (!bjjValid) {
+      throw new Error('signature proof not valid');
+    }
+    await this.validateDIDDocumentAuth(
+      proof.issuerData.id,
+      resolverURL,
+      proof.issuerData.state.value
+    );
+
+    const credStatusType = proof.issuerData.credentialStatus.type;
+    const credStatusResolver = await credStatusResolverRegistry.get(credStatusType);
+    if (!credStatusResolver) {
+      throw new Error(`please register credential status resolver for ${credStatusType} type`);
+    }
+    const credStatus = await credStatusResolver.resolve(proof.issuerData.credentialStatus, {
+      issuerDID: proof.issuerData.id,
+      userDID: userDID
+    });
+    const stateValid = this.validateTreeState(credStatus.issuer);
+    if (!stateValid) {
+      throw new Error('invalid tree state');
```

Comment:
signature proof: invalid tree state of the issuer while checking credential status of singing key